pkg/store/tsdb.go

```diff
@@ -63,21 +73,68 @@ type ReadWriteTSDBStore struct {
 
 // NewTSDBStore creates a new TSDBStore.
 // NOTE: Given lset has to be sorted.
-func NewTSDBStore(logger log.Logger, db TSDBReader, component component.StoreAPI, extLset labels.Labels) *TSDBStore {
+func NewTSDBStore(logger log.Logger, db TSDBReader, component component.StoreAPI, extLset labels.Labels, metricNameFilterEnabled bool) *TSDBStore {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
-	return &TSDBStore{
+
+	var (
+		storeFilter       filter.StoreFilter
+		startFilterUpdate bool
+	)
+
+	storeFilter = filter.AllowAllStoreFilter{}
+	if metricNameFilterEnabled {
+		startFilterUpdate = true
+		storeFilter = filter.NewCuckooMetricNameStoreFilter(1000000) // about 1MB on 64bit machines.
+	}
+
+	st := &TSDBStore{
 		logger:           logger,
 		db:               db,
 		component:        component,
 		extLset:          extLset,
+		storeFilter:      storeFilter,
 		maxBytesPerFrame: RemoteReadFrameLimit,
+		close:            func() {},
 		buffers: sync.Pool{New: func() interface{} {
 			b := make([]byte, 0, initialBufSize)
 			return &b
 		}},
 	}
+
+	if startFilterUpdate {
+		t := time.NewTicker(15 * time.Second)
```

Comment:
Does this mean there could be a 15 seconds of additional delay when a new metric is written to receivers?

Reply:
...yes, for now it seems sensible and simplest approach.